server/connect.go

```diff
@@ -0,0 +1,149 @@
+package main
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/mattermost/mattermost-plugin-msteams/server/store/storemodels"
+	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/pkg/errors"
+)
+
+const (
+	earlyAdopterThreshold = 1000
+)
+
+func (p *Plugin) botSendDirectMessage(userID, message string) error {
+	channel, err := p.apiClient.Channel.GetDirect(userID, p.userID)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get bot DM channel with user_id %s", userID)
+	}
+
+	return p.apiClient.Post.CreatePost(&model.Post{
+		Message:   message,
+		UserId:    p.userID,
+		ChannelId: channel.Id,
+	})
+}
+
+func (p *Plugin) MaybeSendInviteMessage(userID string) (bool, error) {
+	if p.getConfiguration().ConnectedUsersInvitePoolSize == 0 {
+		// connection invites disabled
+		return false, nil
+	}
+
+	var nWhitelisted int
+	var pendingSince time.Time
+	now := time.Now()
+
+	user, err := p.apiClient.User.Get(userID)
+	if err != nil {
+		p.API.LogWarn("Error getting user", "user_id", userID, "error", err.Error())
+		return false, err
+	}
+
+	userInWhitelist, err := p.store.IsUserPresentInWhitelist(user.Id)
+	if err != nil {
+		p.API.LogWarn("Error getting user in whitelist", "user_id", userID, "error", err.Error())
+		return false, err
+	}
+
+	if userInWhitelist {
+		// user already connected
+		return false, nil
+	}
+
+	invitedUser, err := p.store.GetInvitedUser(user.Id)
+	if err != nil {
+		p.API.LogWarn("Error getting user invite", "user_id", userID, "error", err.Error())
+		return false, err
+	}
+
+	if invitedUser != nil {
+		pendingSince = invitedUser.InvitePendingSince
+	} else {
+		moreInvitesAllowed, n, err := p.moreInvitesAllowed()
+		if err != nil {
+			p.API.LogWarn("Error checking invite pool size", "error", err.Error())
+			return false, errors.Wrapf(err, "Error checking invite pool size")
+		}
+
+		if !moreInvitesAllowed {
+			// user not connected, but invite threshold is presently met
+			return false, nil
+		}
+
+		nWhitelisted = n
+	}
+
+	if !p.shouldSendInviteMessage(pendingSince, now, user.GetTimezoneLocation()) {
+		return false, nil
+	}
+
+	if err := p.SendInviteMessage(user, pendingSince, now, nWhitelisted); err != nil {
+		p.API.LogWarn("Error sending connection invite", "error", err.Error())
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (p *Plugin) SendInviteMessage(user *model.User, pendingSince time.Time, currentTime time.Time, nWhitelisted int) error {
+	invitedUser := &storemodels.InvitedUser{ID: user.Id, InvitePendingSince: pendingSince, InviteLastSentAt: currentTime}
+	if invitedUser.InvitePendingSince.IsZero() {
+		invitedUser.InvitePendingSince = currentTime
+	}
+
+	if err := p.store.StoreInvitedUser(invitedUser); err != nil {
+		p.API.LogWarn("Error storing user in invite list", "error", err.Error())
+		return err
+	}
+
+	connectURL := p.GetURL() + "/connect"
+
+	var message string
+	if nWhitelisted < earlyAdopterThreshold {
```

Comment:
Propose we simplify to a single message vs distinguishing (and hard-coding) a threshold for early adopters.

Reply:
0/5, happy to remove, just following https://docs.google.com/document/d/14Wu18J5Ss-rnxlwmFRiMg2Qt4qih3Qo_Q-0bO4rkBP8/edit?usp=sharing. @johndavidlugtu good with you to remove?